### PR TITLE
(PA-1763) Patch ruby CVE 2017-17405

### PR DESCRIPTION
**Pull request by caseywilliams**
Two commits:
- Remove unused ruby 2.3.1, 2.3.3, and 2.4.1 files and configurations (The only currently used versions of ruby in puppet-agent are 2.1.9 and 2.4.2)
- Patch ruby 2.1.9 for CVE [2017-17405](https://www.ruby-lang.org/en/news/2017/12/14/net-ftp-command-injection-cve-2017-17405/). As noted in the patch file, I've included only the fix to Net::FTP here, and left out changes related to version and tests, since some didn't apply cleanly.

### CODE REVIEW

**Comment by caseywilliams** [comment]
jenkins please test this

File: resources/patches/ruby_219/cve-2017-17405.patch [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+	Fix a command injection vulnerability in Net::FTP.
+
+git-svn-id: svn+ssh://ci.ruby-lang.org/ruby/branches/ruby_2_2@61246 b2dd03c8-39d4-4d8f-98ff-823fe69b080e
+---
+ ChangeLog                |   4 +
+ lib/net/ftp.rb           |  10 +-
+ test/net/ftp/test_ftp.rb | 234 +++++++++++++++++++++++++++++++++++++++++++++++
+ version.h                |   2 +-
+ 4 files changed, 244 insertions(+), 6 deletions(-)
```
Comment (caseywilliams) [inline resources/patches/ruby_219/cve-2017-17405.patch]: I kept only the `lib/net/ftp.rb` changes - `version.h` didn't apply cleanly - is that ok?

**Comment by geoffnichols** [comment]
Looks like the Travis failure is because the Jira ticket is set to private. But since CVE 2017-17405 is a public CVE in an upstream component (Ruby), I think we could make PA-1763 a public ticket.

**Comment by caseywilliams** [comment]
jenkins please test this on windows2012r2-64,aix71-POWER

...just in case

**Comment by geoffnichols** [comment]
This is good to merge after update to patch header and commit message.

**Comment by caseywilliams** [comment]
jenkins please test this on windows2012r2-64,solaris10-SPARC

**Comment by caseywilliams** [comment]
I've run this through the ad-hoc pipeline to test AIX and a few more platforms [here](https://jenkins-master-prod-1.delivery.puppetlabs.net/view/puppet-agent/view/ad-hoc/job/platform_puppet-agent-extra_puppet-agent-integration-suite_adhoc-ad_hoc/42/) - I think this is ready for merge unless anyone wants to see more extensive testing.

**Pull request merged**